src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -0,0 +1,4300 @@
+					geo->set_used_shaders( shaderAssignPair.second );
+				}
+				else if( shaderAssignPair.first->is_a( ccl::Light::get_node_type() ) )
+				{
+					ccl::Light *light = static_cast<ccl::Light*>( shaderAssignPair.first );
+					if( shaderAssignPair.second[0] )
+					{
+						ccl::Shader *shader = static_cast<ccl::Shader *>( shaderAssignPair.second[0] );
+						shader->tag_used( m_scene );
+						light->set_shader( shader );
+					}
+					else
+					{
+						light->set_shader( m_scene->default_light );
+					}
+				}
+			}
+			m_shaderAssignPairs.clear();
+
+			// TODO: Optimise
+			for( ccl::Light *light : m_scene->lights )
+			{
+				if( light->get_light_type() == ccl::LIGHT_BACKGROUND )
+				{
+					// Set environment map rotation
+					Imath::M44f transform =  SocketAlgo::getTransform( light->get_tfm() );
+					Imath::Eulerf euler( transform, Imath::Eulerf::Order::XZY );
+
+					for( ccl::ShaderNode *node : light->get_shader()->graph->nodes )
+					{
+						if ( node->type == ccl::EnvironmentTextureNode::node_type )
+						{
+							ccl::EnvironmentTextureNode *env = (ccl::EnvironmentTextureNode *)node;
+							env->tex_mapping.rotation = ccl::make_float3( -euler.x, -euler.y, -euler.z );
+							light->get_shader()->tag_update( m_scene );
+							break;
+						}
+					}
+				}
+			}
+
+			ccl::vector<ccl::Shader *> &shaders = m_scene->shaders;
+			if( nodes.size() + m_numDefaultShaders > shaders.size() )
+			{
+				shaders.resize( m_numDefaultShaders );
+				for( ccl::Node *node : nodes )
+				{
+					ccl::Shader *shader = static_cast<ccl::Shader *>( node );
+					shaders.push_back( shader );
+				}
+				m_scene->shader_manager->tag_update( m_scene, ccl::ShaderManager::SHADER_ADDED );
+				// TODO: Optimise
+				m_scene->background->tag_update( m_scene );
+			}
+			nodes.clear();
+		}
+
+		ccl::Scene *m_scene;
+		int m_numDefaultShaders;
+		typedef tbb::concurrent_hash_map<IECore::MurmurHash, CyclesShaderPtr> Cache;
+		Cache m_cache;
+		CyclesShaderPtr m_defaultSurface;
+		// Need to assign shaders in a deferred manner
+		typedef tbb::concurrent_vector<ShaderAssignPair> ShaderAssignVector;
+		ShaderAssignVector m_shaderAssignPairs;
+
+};
+
+IE_CORE_DECLAREPTR( ShaderCache )
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// CyclesAttributes
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+// Standard Attributes
+IECore::InternedString g_visibilityAttributeName( "visibility" );
+IECore::InternedString g_transformBlurAttributeName( "transformBlur" );
+IECore::InternedString g_transformBlurSegmentsAttributeName( "transformBlurSegments" );
+IECore::InternedString g_deformationBlurAttributeName( "deformationBlur" );
+IECore::InternedString g_deformationBlurSegmentsAttributeName( "deformationBlurSegments" );
+// Cycles Attributes
+IECore::InternedString g_cclVisibilityAttributeName( "ccl:visibility" );
```

Comment:
I decided upon using `ccl:` as the namespace prefix to be in-line with the C++ internal namespace, as well as not interfering with OpenCL's `cl` namespace either, similarly to Gaffer's choice of using `ai:` namespace that Arnold uses from `libai.so`. In contrast the USD schema has dediced to use the `arnold:` namespace instead, would it be worth biting the bullet now and decide to use the `cycles:` prefix to be compatible with USD instead of making code to translate this in `IECoreUSD`? https://github.com/blender/cycles/blob/master/src/hydra/render_delegate.h#L15 looks like Nvidia decided upon `cycles:` namespace like the Tangent HdCycles...

Reply:
Yeah, I think adopting a `cycles:` convention before we make a first official release is a good call. But I'd like to get your code merged in its original form first, and then iterate from there.